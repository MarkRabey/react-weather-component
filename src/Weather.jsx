import React from 'react';

const API_KEY = 'ab3b62af06e932c934c1e7b9de501f00';

export default class Weather extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};
  }

  componentDidMount() {
    const options = {
      enableHighAccuracy: true,
      timeout: 5000,
      maximumAge: 0,
    };

    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(
        pos => {
          this.setState({
            coordinates: pos.coords,
          });
          this.checkLocation();
        },
        () => {
          this.checkLocation();
        },
        options,
      );
    }

    setInterval(() => this.checkLocation(), 10 * 60 * 1000);
  }

  checkLocation() {
    if (!this.state.coordinates) {
      fetch('https://ipinfo.io/json')
        .then(res => res.json())
        .then(ip => {
          this.setState({
            coordinates: {
              latitude: +ip.loc.split(',')[0],
              longitude: +ip.loc.split(',')[1],
            },
          });

          this.fetchForecast();
        });
    } else {
      this.fetchForecast();
    }
  }

  fetchForecast() {
    const {coordinates} = this.state;
    const API_URL = `http://api.openweathermap.org/data/2.5/weather?units=metric&lat=${coordinates.latitude}&lon=${coordinates.longitude}&appid=${API_KEY}`;
    fetch(API_URL)
      .then(c => c.json())
      .then(forecast => {
        this.setState({forecast});
      });
  }

  renderWeather() {
    if (!this.state.forecast) {
      return (
        <div className="weather-container">
          <p>Loading...</p>
        </div>
      );
    }

    return (
      <div className="weather-container">
        <h2>Weather for {this.state.forecast.name}</h2>
        <p>Current temperature: {this.state.forecast.main.temp}&deg;c</p>
        <p>Forecast High: {this.state.forecast.main.temp_max}&deg;c</p>
        <p>Forecast Low: {this.state.forecast.main.temp_min}&deg;c</p>
        <p>Humidity: {this.state.forecast.main.humidity}%</p>
        <p>Wind Speed: {this.state.forecast.wind.speed} km/h</p>
      </div>
    );
  }

  render() {
    return <div>{this.renderWeather()}</div>;
  }
}
