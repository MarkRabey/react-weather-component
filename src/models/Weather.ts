export interface WeatherResponse {
  coord: {lat: number; lon: number};
  weather: WeatherSummary[];
  base: string;
  main: WeatherDetails;
  visibility: number;
  wind: Wind;
  clouds: Clouds;
  dt: number;
  sys: Sys;
  timezone: number;
  id: number;
  name: string;
  cod: number;
}

export interface WeatherSummary {
  id: number;
  main: string;
  description: string;
  icon: string;
}

export interface WeatherDetails {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
  sea_level: number;
  grnd_level: number;
}

export interface Wind {
  speed: number;
  deg: number;
  gust: number;
}

export interface Clouds {
  all: number;
}

export interface Sys {
  type: number;
  id: number;
  country: string;
  sunrise: number;
  sunset: number;
}
