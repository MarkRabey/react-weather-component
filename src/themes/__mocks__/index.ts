import {WeatherStyles} from '../types';

export const mockStyles: WeatherStyles = {
  container: {},
  headingContainer: {},
  heading: {},
  subheading: {},
  inner: {},
  detailsContainer: {},
  details: {},
};

export const defaultLightStyles: WeatherStyles = {
  container: {
    flex: 1,
    borderWidth: 2,
    borderRadius: 4,
    padding: 12,
    backgroundColor: '#f9fafb',
    borderColor: '#e5e7eb',
  },
  inner: {},
  headingContainer: {flex: 1, marginBottom: 6},
  heading: {
    fontFamily: 'sans-serif',
    fontSize: 24,
    fontWeight: 'bold',
    color: '#000000',
  },
  subheading: {fontSize: 12, color: '#8b5cf6'},
  detailsContainer: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: -8,
  },
  details: {color: '#8b5cf6', fontSize: 12},
};

export const defaultMonospaceStyles: WeatherStyles = {
  container: {
    flex: 1,
    borderWidth: 2,
    borderRadius: 4,
    padding: 12,
    backgroundColor: '#f9fafb',
    borderColor: '#e5e7eb',
  },
  inner: {},
  headingContainer: {flex: 1, marginBottom: 6},
  heading: {
    fontFamily: 'monospace',
    fontSize: 24,
    fontWeight: 'bold',
    color: '#000000',
  },
  subheading: {fontSize: 12, color: '#8b5cf6', fontFamily: 'monospace'},
  detailsContainer: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: -8,
  },
  details: {color: '#8b5cf6', fontSize: 12, fontFamily: 'monospace'},
};
