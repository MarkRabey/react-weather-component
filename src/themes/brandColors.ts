export default {
  white: '#ffffff',
  black: '#000000',
  lightText: '#ffffff',
  darkText: '#000000',
  violet: {
    50: '#f5f3ff',
    100: '#ede9fe',
    200: '#ddd6fe',
    300: '#c4b5fd',
    400: '#a78bfa',
    500: '#8b5cf6',
    600: '#7c3aed',
    700: '#6d28d9',
    800: '#5b21b6',
    900: '#4c1d95',
  },
  coolGray: {
    50: '#f9fafb',
    100: '#f3f4f6',
    200: '#e5e7eb',
    300: '#d1d5db',
    400: '#9ca3af',
    500: '#6b7280',
    600: '#4b5563',
    700: '#374151',
    800: '#1f2937',
    900: '#111827',
  },
};
