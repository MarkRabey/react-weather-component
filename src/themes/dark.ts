import {brandColors} from '.';
import {Theme} from './types';

const darkTheme: Theme = {
  name: 'dark',
  containerStyles: {
    backgroundColor: brandColors.coolGray[600],
    borderColor: brandColors.coolGray[600],
  },
  headingStyles: {
    color: brandColors.lightText,
  },
  subheadingStyles: {
    color: brandColors.violet[400],
  },
  detailStyles: {
    color: brandColors.violet[400],
  },
};

export default darkTheme;
