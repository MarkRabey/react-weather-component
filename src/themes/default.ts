import {Theme} from '.';

const defaultTheme: Theme = {
  name: 'default',
  containerStyles: {
    borderWidth: 2,
    borderRadius: 4,
    borderColor: '#333',
    padding: 12,
  },
  headingStyles: {
    fontFamily: 'sans-serif',
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 6,
  },
  detailStyles: {
    fontFamily: 'serif',
    color: '#333',
  },
};

export default defaultTheme;
