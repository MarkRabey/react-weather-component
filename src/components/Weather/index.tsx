import React, {useCallback, useEffect, useState} from 'react';
import {Text, View} from 'react-native';

import {Coordinates} from '../../models/Coordinates';
import {WeatherResponse} from '../../models/Weather';
import {getCoordinates, getWeather} from '../../api';

interface Props {
  apiKey: string;
}

const Weather: React.FC<Props> = ({apiKey}) => {
  const [isLoading, setIsLoading] = useState(true);
  const [forecast, setForecast] = useState<WeatherResponse | null>(null);
  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);

  const getLocation = async () => {
    const coords = await getCoordinates();
    setCoordinates(coords);
  };

  const getForecast = useCallback(async () => {
    if (coordinates) {
      await setIsLoading(true);
      const data = await getWeather(coordinates, apiKey);
      setForecast(data);
      setIsLoading(false);
    }
  }, [apiKey, coordinates]);

  useEffect(() => {
    if (!forecast) {
      if (coordinates) {
        getForecast();
      } else {
        getLocation();
      }
    }
  }, [coordinates, forecast, getForecast]);

  return forecast ? (
    <View>
      {isLoading ? (
        <Text>...Loading</Text>
      ) : (
        <>
          <Text style={{fontSize: 24}}>Weather for {forecast.name}</Text>
          <Text>Weather for {forecast.name}</Text>
          <Text>Current Temperature: {forecast.main.temp}°c</Text>
          <Text>Forecast High: {forecast.main.temp_max}°c</Text>
          <Text>Forecast Low: {forecast.main.temp_min}°c</Text>
          <Text>Humidity: {forecast.main.humidity}%</Text>
          <Text>Wind Speed: {forecast.wind.speed}%</Text>
        </>
      )}
    </View>
  ) : (
    <Text>Something went wrong...</Text>
  );
};

export default Weather;
