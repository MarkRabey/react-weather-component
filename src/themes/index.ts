import brandColors from './brandColors';
import darkTheme from './dark';
import lightTheme from './light';
import monospaceTheme from './monospace';
import monospaceLightTheme from './monospace-light';
import monospaceDarkTheme from './monospace-dark';
export {brandColors};
export const themes = {
  dark: darkTheme,
  light: lightTheme,
  monospace: monospaceTheme,
  'monospace-light': monospaceLightTheme,
  'monospace-dark': monospaceDarkTheme,
};

export type SystemTheme = keyof typeof themes;
