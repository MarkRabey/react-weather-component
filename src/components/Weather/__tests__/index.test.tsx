import React from 'react';
import renderer from 'react-test-renderer';

import Weather from '..';
import {Theme, WeatherStyles} from '../../../themes/types';

// console.log({hooks});
const mockStyles: WeatherStyles = {
  container: {},
  headingContainer: {},
  heading: {},
  subheading: {},
  inner: {},
  detailsContainer: {},
  details: {},
};

const mockSelectedTheme: Theme = {name: 'testing'};
jest.mock('../../../hooks/useStyles', () => ({
  useStyles: () => ({
    styles: mockStyles,
    selectedTheme: mockSelectedTheme,
  }),
}));

describe('WeatherComponent', () => {
  test('renders', () => {
    const result = renderer.create(<Weather apiKey="" />).toJSON();
    expect(result).toMatchSnapshot();
  });
});
