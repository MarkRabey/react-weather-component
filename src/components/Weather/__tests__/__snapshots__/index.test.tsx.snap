// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherComponent renders 1`] = `
<div
  className="css-view-1dbjc4n"
>
  <div
    className="css-text-901oao"
    dir="auto"
  >
    <div
      aria-valuemax={1}
      aria-valuemin={0}
      className="css-view-1dbjc4n"
      role="progressbar"
      style={
        Object {
          "WebkitAlignItems": "center",
          "WebkitBoxAlign": "center",
          "WebkitBoxPack": "center",
          "WebkitJustifyContent": "center",
          "alignItems": "center",
          "display": "inline-flex",
          "justifyContent": "center",
          "msFlexAlign": "center",
          "msFlexPack": "center",
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-animationKeyframes-127358a"
        style={
          Object {
            "WebkitAnimationDuration": "0.75s",
            "WebkitAnimationIterationCount": "infinite",
            "WebkitAnimationTimingFunction": "linear",
            "animationDuration": "0.75s",
            "animationIterationCount": "infinite",
            "animationTimingFunction": "linear",
            "display": "inline-flex",
            "height": "36px",
            "width": "36px",
          }
        }
      >
        <svg
          height="100%"
          viewBox="0 0 32 32"
          width="100%"
        >
          <circle
            cx="16"
            cy="16"
            fill="none"
            r="14"
            strokeWidth="4"
            style={
              Object {
                "opacity": 0.2,
                "stroke": "#1976D2",
              }
            }
          />
          <circle
            cx="16"
            cy="16"
            fill="none"
            r="14"
            strokeWidth="4"
            style={
              Object {
                "stroke": "#1976D2",
                "strokeDasharray": 80,
                "strokeDashoffset": 60,
              }
            }
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
